src/runtime_safe_heap.js

```diff
@@ -8,91 +8,39 @@
 #error "should only be inclded in SAFE_HEAP mode"
 #endif
 
-/** @param {number|boolean=} isFloat */
-function getSafeHeapType(bytes, isFloat) {
-  switch (bytes) {
-    case 1: return 'i8';
-    case 2: return 'i16';
-    case 4: return isFloat ? 'float' : 'i32';
-    case 8: return isFloat ? 'double' : 'i64';
-    default: abort(`getSafeHeapType() invalid bytes=${bytes}`);
-  }
-}
-
 #if SAFE_HEAP_LOG
 var SAFE_HEAP_COUNTER = 0;
 #endif
 
-/** @param {number|boolean=} isFloat */
-function SAFE_HEAP_STORE(dest, value, bytes, isFloat) {
+function SAFE_HEAP_INDEX(arr, idx, log) {
 #if CAN_ADDRESS_2GB
-  dest >>>= 0;
+  idx >>>= 0;
 #endif
+  const bytes = arr.BYTES_PER_ELEMENT;
+  const dest = idx * bytes;
 #if SAFE_HEAP_LOG
-  dbg('SAFE_HEAP store: ' + [dest, value, bytes, isFloat, SAFE_HEAP_COUNTER++]);
-#endif
-  if (dest <= 0) abort(`segmentation fault storing ${bytes} bytes to address ${dest}`);
-#if SAFE_HEAP == 1
-  if (dest % bytes !== 0) abort(`alignment error storing to address ${dest}, which was expected to be aligned to a multiple of ${bytes}`);
-#else
-  if (dest % bytes !== 0) warnOnce(`alignment error in a memory store operation, alignment was a multiple of ${(((dest ^ (dest-1)) >> 1) + 1)}, but was was expected to be aligned to a multiple of ${bytes}`);
+  dbg(`SAFE_HEAP ${log}: ${[arr.constructor.name, idx, SAFE_HEAP_COUNTER++]}`);
 #endif
+  if (idx <= 0) abort(`segmentation fault ${log} ${bytes} bytes from address ${dest}`);
 #if EXIT_RUNTIME
   if (runtimeInitialized && !runtimeExited) {
 #else
   if (runtimeInitialized) {
 #endif
     var brk = _sbrk(0);
-    if (dest + bytes > brk) abort(`segmentation fault, exceeded the top of the available dynamic heap when storing ${bytes} bytes to address ${dest}. DYNAMICTOP=${brk}`);
+    if (dest + bytes > brk) abort(`segmentation fault, exceeded the top of the available dynamic heap when ${log} ${bytes} bytes from address ${dest}. DYNAMICTOP=${brk}`);
```

Comment:
I guess this can be rephrased now to something like `when attempting to ${log} ${bytes} bytes`

Reply:
(otherwise the grammer looks wrong I think)

Reply:
(sorry ignore that.. I see that actions do are of the `ing` form).